### PR TITLE
[WFCORE-3305][WFCORE-3425] Add more runtime operations to the Elytron key-store resource for advanced KeyStore manipulation

### PR DESCRIPTION
**Pull request by fjuma**
This PR adds more runtime-only management operations to the Elytron subsystem key-store resource for advanced KeyStore manipulation. The new operations are as follows:

* **generate-key-pair**: This will generate a key pair and wrap the resulting public key in a self-signed X.509 certificate. The generated private key and self-signed certificate will be added to the KeyStore.

* **generate-certificate-signing-request**: This will generate a PKCS #10 certificate signing request using a PrivateKeyEntry from the KeyStore. The generated certificate signing request will be output to a file.

* **import-certificate**: This will import a certificate or certificate chain from a file into an entry in the KeyStore.

* **export-certificate**: This will export a certificate from an entry in the KeyStore to a file.

* **change-alias**: This will move an existing KeyStore entry to a new alias.

After executing the generate-key-pair, import-certificate, or change-alias operations, executing the store operation will persist the resulting changes to the file that backs the KeyStore.

This PR also adds a commit to fix WFCORE-3425. This ensures that for a file-backed key-store, if the file doesn’t exist and wasn’t flagged as required, then the file will get created properly when executing the store operation.

https://issues.jboss.org/browse/WFCORE-3305
https://issues.jboss.org/browse/WFCORE-3425 

Documentation PR: https://github.com/wildfly/wildfly/pull/10830

### CODE REVIEW

**Comment by fjuma** [comment]
Added a few more tests for the _import-certificate_ operation as discussed with Martin Choma.

**Comment by fjuma** [comment]
retest this please

File: elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+                    }
+                }
+
+                SelfSignedX509CertificateAndSigningKey certAndKey = certAndKeyBuilder.build();
+                final PrivateKey privateKey = certAndKey.getSigningKey();
+                final X509Certificate[] certChain = new X509Certificate[1];
+                certChain[0] = certAndKey.getSelfSignedCertificate();
+                keyStore.setKeyEntry(alias, privateKey, keyPassword, certChain);
+            } catch (KeyStoreException | IllegalArgumentException | DateTimeException e) {
+                throw new OperationFailedException(e);
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java]: Do all of these indicate user errors? If any don't, throw a runtime exception.

File: elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+
+                PKCS10CertificateSigningRequest csr = csrBuilder.build();
+                try (FileOutputStream fos = new FileOutputStream(resolvedPath)) {
+                    fos.write(csr.getPem());
+                }
+            } catch (Exception e) {
+                if (e instanceof OperationFailedException) {
+                    throw (OperationFailedException) e;
+                } else {
+                    throw new OperationFailedException(e);
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java]: Don't throw OFE unless it's a user mistake. If not use a runtime exception.

File: elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+                            }
+                        }
+                    }
+                    keyStore.setCertificateEntry(alias, trustedCertificate);
+                }
+            } catch (Exception e) {
+                if (e instanceof OperationFailedException) {
+                    throw (OperationFailedException) e;
+                } else {
+                    throw new OperationFailedException(e);
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java]: Same thing.

File: elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+                        fos.write(pemCertificate.toArray());
+                    } else {
+                        fos.write(certificate.getEncoded());
+                    }
+                }
+            } catch (Exception e) {
+                if (e instanceof OperationFailedException) {
+                    throw (OperationFailedException) e;
+                } else {
+                    throw new OperationFailedException(e);
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java]: One more time. :)

File: elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java [hunk 5]
```diff
@@ -0,0 +1,684 @@
...
+                }
+                keyStore.setEntry(newAlias, entry, passwordProtection);
+                if (keyStore.containsAlias(alias)) {
+                    keyStore.deleteEntry(alias);
+                }
+            } catch (Exception e) {
+                if (e instanceof OperationFailedException) {
+                    throw (OperationFailedException) e;
+                } else {
+                    throw new OperationFailedException(e);
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java]: etc

File: elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+                }
+            }
+        }
+    }
+
+    private static char[] resolveKeyPassword(final KeyStoreService keyStoreService, final ExceptionSupplier<CredentialSource, Exception> credentialSourceSupplier) throws OperationFailedException {
+        try {
+            return keyStoreService.resolveKeyPassword(credentialSourceSupplier);
+        } catch (Exception e) {
+            throw new OperationFailedException(e);
```
Comment (bstansberry) [inline elytron/src/main/java/org/wildfly/extension/elytron/AdvancedModifiableKeyStoreDecorator.java]: etc

**Comment by bstansberry** [comment]
This should get an approval by someone more familiar with elytron and with the requirements. It also needs to be out of the analysis phase so we know those requirements are set. :)  Seems ok to me other than my notes about exception handling.

If there's additional external testing of this going on, please comment noting that and comment when that's done.

**Comment by fjuma** [comment]
I've updated the exception handling.

The two test failures currently occurring on Windows (```KeyStoresTestCase.testExportCertificatePem``` and ```KeyStoresTestCase.testGenerateCertificateSigningRequest```) should be fixed once Core has been upgraded to Elytron 1.2.0.Beta12.

**Comment by fjuma** [comment]
Just a note that Martin Choma has moved this issue out of the analysis phase now so the requirements are set.

**Comment by bstansberry** [comment]
retest this please

**Comment by fjuma** [comment]
retest this please

**Comment by bstansberry** [comment]
@fjuma What's the status here? Monday is about it for getting this into WF 12 and no one from the security team has reviewed this and there's no word on whether QE's ok with it. This also blocks #3055.

**Comment by fjuma** [comment]
@mchoma Please add a comment here on the test status.

@darranl Please review.

Thanks.

**Comment by mchoma** [comment]
I have marked EAP7-650 as Test Development done. There are no blockers on this. I am OK with getting this into WF12.

File: elytron/src/main/java/org/wildfly/extension/elytron/KeyStoreService.java [hunk 7]
```diff
@@ -271,6 +266,21 @@ boolean isModified() {
         return trackingKeyStore.isModified();
     }
 
+    char[] resolveKeyPassword(final ExceptionSupplier<CredentialSource, Exception> keyPasswordCredentialSourceSupplier) throws Exception {
```
Comment (darranl) [inline elytron/src/main/java/org/wildfly/extension/elytron/KeyStoreService.java]: Not related to this PR but we also want at some point to look at KeyStore password re-use for the KeyManager resource as well.

**Comment by darranl** [comment]
Do we have any information about the new binary files so if anyone is debugging or needs to recreate them they know what they need?

**Comment by fjuma** [comment]
@darranl I've added a readme file with information about the new binary files.

**Comment by darranl** [comment]
Thanks @fjuma looks good.

**Comment by bstansberry** [comment]
retest this please

**Comment by bstansberry** [comment]
@fjuma I resolved the two WFCORE's mentioned in the title but I'll leave any EAP7 etc manipulation to you. :)

**Pull request merged**